File: windows/src/AdBlocker/Core/Updates.cs
using System.Security.Cryptography;
using System.Text.Json;
using AdBlocker.Config;

namespace AdBlocker.Core;

/// <summary>A release on the project's GitHub page.</summary>
public sealed record Release(string Version, string ExeUrl, long ExeSize, string? SumsUrl, string PageUrl);

/// <summary>
/// Looks for a newer AdBlocker on the project's GitHub releases page and downloads it.
/// The executable is not code-signed, so the SHA-256 published with the release is checked;
/// that catches a broken download, and beyond it this trusts GitHub like a manual download does.
/// </summary>
public static class Updates
{
    public const string LatestReleaseApi = "https://api.github.com/repos/teykaijun/AdBlockerPlugin/releases/latest";
    public const string ExeName = "adblocker.exe";
    public const string SumsName = "SHA256SUMS.txt";
    private const long MaxBytes = 100 * 1024 * 1024;

    /// <summary>Reads GitHub's "latest release" answer. Null if it holds nothing to install.</summary>
    public static Release? ParseRelease(string json)
    {
        using var document = JsonDocument.Parse(json);
        var root = document.RootElement;
        if (root.ValueKind != JsonValueKind.Object) return null;
        if (Flag(root, "draft") || Flag(root, "prerelease")) return null;

        var version = Text(root, "tag_name").TrimStart('v', 'V').Trim();
        if (version.Length == 0) return null;
        if (!root.TryGetProperty("assets", out var assets) || assets.ValueKind != JsonValueKind.Array) return null;

        string? exeUrl = null, sumsUrl = null;
        long exeSize = 0;
        foreach (var asset in assets.EnumerateArray())
        {
            var url = Text(asset, "browser_download_url");
            if (!url.StartsWith("https://", StringComparison.OrdinalIgnoreCase)) continue;
            switch (Text(asset, "name"))
            {
                case ExeName:
                    exeUrl = url;
                    exeSize = asset.TryGetProperty("size", out var size) && size.TryGetInt64(out var value) ? value : 0;
                    break;
                case SumsName:
                    sumsUrl = url;
                    break;
            }
        }
        return exeUrl is null ? null : new Release(version, exeUrl, exeSize, sumsUrl, Text(root, "html_url"));
    }

    /// <summary>Whether <paramref name="latest"/> ("1.4.1" or "v1.4.1") is a higher version than <paramref name="current"/>.</summary>
    public static bool IsNewer(string latest, string current)
    {
        var a = Parts(latest);
        var b = Parts(current);
        for (var i = 0; i < Math.Max(a.Length, b.Length); i++)
        {
            var left = i < a.Length ? a[i] : 0;
            var right = i < b.Length ? b[i] : 0;
            if (left != right) return left > right;
        }
        return false;
    }

    /// <summary>The SHA-256 recorded for <paramref name="fileName"/> in a SHA256SUMS.txt.</summary>
    public static string? ExpectedHash(string sums, string fileName)
    {
        foreach (var line in sums.Split('\n'))
        {
            var parts = line.Trim().Split((char[]?)null, 2, StringSplitOptions.RemoveEmptyEntries);
            if (parts.Length != 2 || parts[0].Length != 64) continue;
            if (parts[1].TrimStart('*').Trim() == fileName) return parts[0].ToLowerInvariant();
        }
        return null;
    }

    /// <summary>The latest release, or null when this build is already up to date.</summary>
    public static async Task<Release?> CheckAsync(HttpClient http, string currentVersion, CancellationToken cancellationToken)
    {
        using var request = new HttpRequestMessage(HttpMethod.Get, LatestReleaseApi);
        request.Headers.Add("Accept", "application/vnd.github+json");
        using var response = await http.SendAsync(request, cancellationToken);
        response.EnsureSuccessStatusCode();
        var release = ParseRelease(await response.Content.ReadAsStringAsync(cancellationToken));
        return release is not null && IsNewer(release.Version, currentVersion) ? release : null;
    }

    /// <summary>Downloads the release's executable into <paramref name="directory"/> and checks its SHA-256.</summary>
    public static async Task<string> DownloadAsync(
        Release release,
        HttpClient http,
        string directory,
        Action<int>? progress,
        CancellationToken cancellationToken)
    {
        if (release.SumsUrl is null) throw new UserError("The release has no checksum file to check the download against.");
        var expected = ExpectedHash(await http.GetStringAsync(release.SumsUrl, cancellationToken), ExeName)
            ?? throw new UserError($"The release's {SumsName} does not list {ExeName}.");

        Directory.CreateDirectory(directory);
        var file = Path.Combine(directory, $"adblocker-{release.Version}.exe");
        using (var response = await http.GetAsync(release.ExeUrl, HttpCompletionOption.ResponseHeadersRead, cancellationToken))
        {
            response.EnsureSuccessStatusCode();
            var total = release.ExeSize > 0 ? release.ExeSize : response.Content.Headers.ContentLength ?? 0;
            await using var input = await response.Content.ReadAsStreamAsync(cancellationToken);
            await using var output = File.Create(file);
            var buffer = new byte[81_920];
            long written = 0;
            var reported = -1;
            int read;
            while ((read = await input.ReadAsync(buffer, cancellationToken)) > 0)
            {
                written += read;
                if (written > MaxBytes) throw new UserError("The download is unexpectedly large.");
                await output.WriteAsync(buffer.AsMemory(0, read), cancellationToken);
                var percent = total > 0 ? (int)Math.Clamp(written * 100 / total, 0, 100) : 0;
                if (percent != reported)
                {
                    reported = percent;
                    progress?.Invoke(percent);
                }
            }
        }

        string actual;
        await using (var written = File.OpenRead(file))
        {
            actual = Convert.ToHexStringLower(await SHA256.HashDataAsync(written, cancellationToken));
        }
        if (!actual.Equals(expected, StringComparison.OrdinalIgnoreCase))
        {
            File.Delete(file);
            throw new UserError("The download does not match the checksum published with the release.");
        }
        return file;
    }

    /// <summary>Records what a check found, so "adblocker status" can mention it without going online.</summary>
    public static void Save(AppPaths paths, Release? release)
    {
        var state = new UpdateState { CheckedAt = DateTimeOffset.UtcNow, Version = release?.Version, PageUrl = release?.PageUrl };
        AtomicFile.WriteAllText(paths.UpdateFile, JsonSerializer.Serialize(state, AppJson.Default.UpdateState));
    }

    public static UpdateState? Read(AppPaths paths)
    {
        try
        {
            return File.Exists(paths.UpdateFile)
                ? JsonSerializer.Deserialize(File.ReadAllText(paths.UpdateFile), AppJson.Default.UpdateState)
                : null;
        }
        catch (Exception e) when (e is JsonException or IOException or UnauthorizedAccessException)
        {
            return null;
        }
    }

    private static int[] Parts(string version) => version.Trim().TrimStart('v', 'V')
        .Split(['.', '-', '+'])
        .Select(part => int.TryParse(part, out var value) ? value : 0)
        .ToArray();

    private static bool Flag(JsonElement element, string name) =>
        element.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.True;

    private static string Text(JsonElement element, string name) =>
        element.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.String ? value.GetString() ?? "" : "";
}

File: android/app/src/main/java/com/teykaijun/adblocker/update/Updates.kt
package com.teykaijun.adblocker.update

import java.io.File
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.security.MessageDigest
import org.json.JSONObject

/** A release on the project's GitHub page. */
data class Release(
    val version: String,
    val apkUrl: String,
    val apkSize: Long,
    val sumsUrl: String?,
    val pageUrl: String,
)

/**
 * Looks for a newer AdBlocker on the project's GitHub releases page and downloads it.
 *
 * Android only replaces an app with a build signed by the same key, and checks that itself,
 * so the worst a wrong download can do is fail to install. The SHA-256 from the release is
 * checked first anyway, to catch a truncated or corrupted file early.
 */
object Updates {
    const val LATEST_RELEASE_API = "https://api.github.com/repos/teykaijun/AdBlockerPlugin/releases/latest"
    const val APK_NAME = "AdBlocker.apk"
    const val SUMS_NAME = "SHA256SUMS.txt"
    private const val TIMEOUT_MS = 20_000
    private const val MAX_APK_BYTES = 100L * 1024 * 1024

    /** Reads GitHub's "latest release" answer. Null if it has no APK to install. */
    fun parseRelease(json: String): Release? {
        val release = JSONObject(json)
        if (release.optBoolean("draft") || release.optBoolean("prerelease")) return null
        val version = release.optString("tag_name").removePrefix("v").trim()
        if (version.isEmpty()) return null

        val assets = release.optJSONArray("assets") ?: return null
        var apk: JSONObject? = null
        var sums: JSONObject? = null
        for (i in 0 until assets.length()) {
            val asset = assets.getJSONObject(i)
            when (asset.optString("name")) {
                APK_NAME -> apk = asset
                SUMS_NAME -> sums = asset
            }
        }
        val apkUrl = apk?.optString("browser_download_url").orEmpty()
        if (!apkUrl.startsWith("https://")) return null
        return Release(
            version = version,
            apkUrl = apkUrl,
            apkSize = apk?.optLong("size") ?: 0,
            sumsUrl = sums?.optString("browser_download_url")?.takeIf { it.startsWith("https://") },
            pageUrl = release.optString("html_url"),
        )
    }

    /** Whether `latest` ("1.4.1" or "v1.4.1") is a higher version number than `current`. */
    fun isNewer(latest: String, current: String): Boolean {
        val a = parts(latest)
        val b = parts(current)
        for (i in 0 until maxOf(a.size, b.size)) {
            val left = a.getOrElse(i) { 0 }
            val right = b.getOrElse(i) { 0 }
            if (left != right) return left > right
        }
        return false
    }

    /** The SHA-256 recorded for `fileName` in a SHA256SUMS.txt, or null if it isn't listed. */
    fun expectedHash(sums: String, fileName: String): String? = sums.lineSequence()
        .mapNotNull { line ->
            val parts = line.trim().split(Regex("""\s+"""), limit = 2)
            if (parts.size == 2 && parts[1].removePrefix("*").trim() == fileName) parts[0].lowercase() else null
        }
        .firstOrNull { it.length == 64 }

    fun sha256(file: File): String {
        val digest = MessageDigest.getInstance("SHA-256")
        file.inputStream().use { input ->
            val buffer = ByteArray(64 * 1024)
            while (true) {
                val read = input.read(buffer)
                if (read < 0) break
                digest.update(buffer, 0, read)
            }
        }
        return digest.digest().joinToString("") { "%02x".format(it) }
    }

    /** The latest release, or null if this build is already up to date. */
    fun check(currentVersion: String): Release? {
        val release = parseRelease(fetch(LATEST_RELEASE_API)) ?: return null
        return release.takeIf { isNewer(it.version, currentVersion) }
    }

    /**
     * Downloads the release's APK into `directory` and checks its SHA-256 against the release.
     * Throws [IOException] if anything does not add up.
     */
    fun download(release: Release, directory: File, onProgress: (Int) -> Unit): File {
        directory.mkdirs()
        directory.listFiles()?.forEach { it.delete() }
        val apk = File(directory, "AdBlocker-${release.version}.apk")
        val expected = release.sumsUrl?.let { expectedHash(fetch(it), APK_NAME) }
            ?: throw IOException("The release has no checksum to verify the download against.")

        open(release.apkUrl) { connection ->
            val total = if (release.apkSize > 0) release.apkSize else connection.contentLengthLong
            connection.inputStream.use { input ->
                apk.outputStream().use { output ->
                    val buffer = ByteArray(64 * 1024)
                    var written = 0L
                    var reported = -1
                    while (true) {
                        val read = input.read(buffer)
                        if (read < 0) break
                        written += read
                        if (written > MAX_APK_BYTES) throw IOException("The download is unexpectedly large.")
                        output.write(buffer, 0, read)
                        val percent = if (total > 0) ((written * 100) / total).toInt().coerceIn(0, 100) else 0
                        if (percent != reported) {
                            reported = percent
                            onProgress(percent)
                        }
                    }
                }
            }
        }

        val actual = sha256(apk)
        if (!actual.equals(expected, ignoreCase = true)) {
            apk.delete()
            throw IOException("The download does not match the checksum published with the release.")
        }
        return apk
    }

    private fun parts(version: String) =
        version.trim().removePrefix("v").split('.', '-', '+').map { it.toIntOrNull() ?: 0 }

    private fun fetch(url: String): String {
        var text = ""
        open(url) { text = it.inputStream.bufferedReader().use { reader -> reader.readText() } }
        return text
    }

    private fun open(url: String, use: (HttpURLConnection) -> Unit) {
        val parsed = URL(url)
        if (parsed.protocol != "https") throw IOException("Only https:// addresses are allowed")
        val connection = parsed.openConnection() as HttpURLConnection
        try {
            connection.connectTimeout = TIMEOUT_MS
            connection.readTimeout = TIMEOUT_MS
            connection.setRequestProperty("Accept", "application/vnd.github+json")
            connection.setRequestProperty("User-Agent", "AdBlocker-Android")
            if (connection.responseCode !in 200..299) throw IOException("GitHub answered HTTP ${connection.responseCode}")
            use(connection)
        } finally {
            connection.disconnect()
        }
    }
}
